[release/9.0-staging] Fix to #35239 - EF9: SaveChanges() is significantly slower in .NET9 vs. .NET8 when using .ToJson() Mapping vs. PostgreSQL Legacy POCO mapping

maumar opened the pull request:
Fixes #35239
Safer version of https://github.com/dotnet/efcore/pull/35326

**Description**

EF9 introduced a change in how we construct ValueComparers for some of our types (specifically collection of scalars/references), in preparation for AOT work. The way the change was implemented may cause a severe performance regression during SaveChanges operation involving multiple entities using collections of primitives (one of our highly requested features). 

**Customer impact**

Customers performing data manipulation operations on entities with collections of primitives may experience significant performance regressions. This may also happen when no data has been changed, but sufficiently large entity graph has been loaded into change tracker. There is no workaround for this issue, apart from changing the model to not use primitive collections (which is unacceptable for majority of customers)

**How found**

Multiple customer reports on EF 9

**Regression**

Yes, from EF8. Note: this is a perf regression only, not a functional regression.

**Testing**

Ad hoc performance test using BenchmarkDotNet. Functional testing already covered by existing tests.

**Risk**

Low. The patch fix has been limited in scope to reduce the risk. Changes should only affect models with primitive collections. Added quirks just to be sure.

**Perf numbers**

Over 20x regression between 8 and 9 for the tested scenario. 

8.0.11

| Method          | Mean     | Error   | StdDev  |
|---------------- |---------:|--------:|--------:|
| SaveChangesTest | 172.1 ms | 1.78 ms | 1.58 ms |


9.0

| Method          | Mean    | Error    | StdDev   |
|---------------- |--------:|---------:|---------:|
| SaveChangesTest | 5.487 s | 0.0621 s | 0.0551 s |


9.0.2 (with fix)

| Method          | Mean     | Error   | StdDev  |
|---------------- |---------:|--------:|--------:|
| SaveChangesTest | 179.8 ms | 1.71 ms | 1.43 ms |

AndriySvyryd commented on src/EFCore.Cosmos/ChangeTracking/Internal/StringDictionaryComparer.cs:
```diff
@@ -68,9 +94,23 @@ private static Expression<Func<object, int>> GetHashCodeLambda(ValueComparer ele
+                    GetHashCodeMethod,
+                    Expression.Convert(
+                        prm,
+                        typeof(IEnumerable)),
```
You don't need this cast as the first thing `GetHashCode` does is to cast it again.

maumar replied:
will keep to be in line with the original code - HashCode method requires IEnumerable as source, so we should make sure the types match, for the purpose of constructing the lambda

AndriySvyryd commented on src/EFCore/ChangeTracking/ListOfNullableValueTypesComparer.cs:
```diff
@@ -67,10 +82,23 @@ ValueComparer IInfrastructure<ValueComparer>.Instance
+        if (!UseOldBehavior35239)
+        {
+            //(a, b) => Compare(a, b, elementComparer.Equals)
+            return Expression.Lambda<Func<IEnumerable<TElement?>?, IEnumerable<TElement?>?, bool>>(
```
Consider the case of `new long?[]{ (int)0 }`

maumar replied:
added checks for LoVT and LoNVT

maumar commented:
@AndriySvyryd update: added changes to the cosmos StringDictionaryComparer. It does support nested collections and the legacy functions can be invoked (rather than just being constructed like in relational). I added targeted fix similar to what I initially did for main.

AndriySvyryd commented:
Looks good. Can you add the servicing template?

maumar commented:
/azp run
